test/hotspot/gtest/runtime/test_atomic.cpp

```diff
@@ -0,0 +1,695 @@
+
+  explicit TranslatedAtomicTestObject1(int value)
+    : _value(value) {}
+};
+
+template<>
+struct PrimitiveConversions::Translate<TranslatedAtomicTestObject1>
+  : public std::true_type
+{
+  using Value = TranslatedAtomicTestObject1;
+  using Decayed = int;
+
+  static Decayed decay(Value x) { return x._value; }
+  static Value recover(Decayed x) { return Value(x); }
+};
+
+struct TranslatedAtomicTestObject2 {
+  TranslatedAtomicTestObject1 _value;
+
+  static constexpr int DefaultObject1Value = 3;
+
+  TranslatedAtomicTestObject2()
+    : TranslatedAtomicTestObject2(TranslatedAtomicTestObject1(DefaultObject1Value))
+  {}
+
+  explicit TranslatedAtomicTestObject2(TranslatedAtomicTestObject1 value)
+    : _value(value) {}
+};
+
+template<>
+struct PrimitiveConversions::Translate<TranslatedAtomicTestObject2>
+  : public std::true_type
+{
+  using Value = TranslatedAtomicTestObject2;
+  using Decayed = TranslatedAtomicTestObject1;
+
+  static Decayed decay(Value x) { return x._value; }
+  static Value recover(Decayed x) { return Value(x); }
+};
+
+struct TranslatedAtomicByteObject {
+  uint8_t _value;
+
+  // NOT default constructible.
+
+  explicit TranslatedAtomicByteObject(uint8_t value = 0) : _value(value) {}
+};
+
+template<>
+struct PrimitiveConversions::Translate<TranslatedAtomicByteObject>
+  : public std::true_type
+{
+  using Value = TranslatedAtomicByteObject;
+  using Decayed = uint8_t;
+
+  static Decayed decay(Value x) { return x._value; }
+  static Value recover(Decayed x) { return Value(x); }
+};
+
+// Test whether Atomic<T> has fetch_then_set().
+// Note: This is intentionally a different implementation from what is used
+// by the atomic translated type to decide whether to provide fetch_then_set().
+// The intent is to make related testing non-tautological.
+// The two implementations must agree; it's a bug if they don't.
+template<typename T>
+class AtomicTypeHasFetchThenSet {
+  template<typename U,
+           typename AU = Atomic<U>,
+           typename = decltype(declval<AU>().fetch_then_set(declval<U>()))>
+  static char* test(int);
+
+  template<typename> static char test(...);
+
+  using test_type = decltype(test<T>(0));
+
+public:
+  static constexpr bool value = std::is_pointer_v<test_type>;
+};
+
+// Unit tests for AtomicTypeHasFetchThenSet.
+static_assert(AtomicTypeHasFetchThenSet<int>::value);
+static_assert(AtomicTypeHasFetchThenSet<int*>::value);
+static_assert(AtomicTypeHasFetchThenSet<TranslatedAtomicTestObject1>::value);
+static_assert(AtomicTypeHasFetchThenSet<TranslatedAtomicTestObject2>::value);
+static_assert(!AtomicTypeHasFetchThenSet<uint8_t>::value);
+
+// Verify translated byte type *doesn't* have fetch_then_set.
+static_assert(!AtomicTypeHasFetchThenSet<TranslatedAtomicByteObject>::value);
+
+// Verify that explicit instantiation doesn't attempt to reference the
+// non-existent fetch_then_set of the atomic decayed type.
+template class AtomicImpl::Atomic<TranslatedAtomicByteObject>;
```

Comment:
Preexisting: 
Why do we not suport exchange on 8 and 16 byte sized types? 

Especially given that we have `AtomicAccess::XchgUsingCmpxchg` and do support `cmpxchg`.